config/src/main/java/com/typesafe/config/impl/Parser.java

```diff
@@ -1063,6 +1063,7 @@ private static Path parsePathExpression(Iterator<Token> expression,
     private static Path parsePathExpression(Iterator<Token> expression,
             ConfigOrigin origin, String originalText) {
         // each builder in "buf" is an element in the path.
+        ArrayList<Token> pathTokens = new ArrayList<Token>();
```

Comment:
It feels like a somewhat messy mixing of concerns to make `Path` and `PathBuilder` keep the tokens around; we do use `Path` in a performance-sensitive context when we parse paths passed to the getters on `Config`, like `Config.getInt`. That's why there's all this fast-path path-parsing stuff. But, even ignoring performance I think it's a little messy. What I'd suggest is to factor this code so there's a version that returns a `Path` and a version that returns a `ConfigNodeKey`, and then a private function which maybe takes the `pathTokens` array list which can be null. So:

```
private static Path parsePathExpression(Iterator<Token> expression, ConfigOrigin origin, String originalText, ArrayList<Token> pathTokens) {
   ...
   if (pathTokens != null) pathTokens.add(t)
   ...
}
```

Then you can have the functions which call that either provide the `ArrayList` or provide null, depending on whether they want to end up with a `Path` or `ConfigNodeKey`

It would be fine to skip the `speculativeFastParsePath` codepath for `ConfigNodeKey`, and always use the slow `parsePathExpression` when we need the tokens. The fast path matters only for that `Config.getInt` case.